### PR TITLE
[native_assets] Add support for link hooks

### PR DESCRIPTION
**Pull request by dcharkes**
This PR adds support invoking `link.dart` hooks.

Link hooks can add new assets. Link hooks can transform assets sent to link hook from build hooks.

This PR does not yet add support for getting tree-shake information in the link hooks. This is pending on defining the `resources.json` format (https://github.com/dart-lang/sdk/issues/55494).

Issue:

* https://github.com/flutter/flutter/issues/146263

## Implementation considerations

The build hooks could be run before Dart compilation and the link hooks after Dart compilation. (This is how it's done in Dart standalone.) However, due to the way the `Target`s are set up, this would require two targets and serializing and deserializing the `BuildResult` in between these. This would lead to more code but no benefits. Currently there is nothing that mandates running build hooks before Dart compilation.

## Testing

* The unit tests verify that the native_assets_builder `link` and `linkDryRun` would be invoked with help of the existing fake.
* The native assets integration test now also invokes an FFI call of a package that adds the asset during the link hook instead of the build hook.
  * In order to keep coverage of the `flutter create --template=package_ffi`, `flutter create` is still run and the extra dependency is added and an extra ffi call is added. (Open to alternative suggestions.)

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[Data Driven Fixes]: https://github.com/flutter/flutter/wiki/Data-driven-Fixes

### CODE REVIEW

**Comment by dcharkes** [comment]
@christopherfujino Is there a way to allow arbitrary git dependencies on the CI here?
(If I make a PR upstream to support a PR here, having to merge and publish that upstream dependency before being able to run the CI here means its likely something will be forgotten upstream and we end up having to do more than one PR upstream.)

**Comment by christopherfujino** [comment]
> @christopherfujino Is there a way to allow arbitrary git dependencies on the CI here? (If I make a PR upstream to support a PR here, having to merge and publish that upstream dependency before being able to run the CI here means its likely something will be forgotten upstream and we end up having to do more than one PR upstream.)

You mean, you want to run presubmit tests against the version of a pub package on some random branch on the internet? You should be able to do `dependency_overrides` with a git dependency in a pubspec.yaml. You may get an analyzer error, but I would expect everything else to Just Work:tm:

**Comment by dcharkes** [comment]
> > @christopherfujino Is there a way to allow arbitrary git dependencies on the CI here? (If I make a PR upstream to support a PR here, having to merge and publish that upstream dependency before being able to run the CI here means its likely something will be forgotten upstream and we end up having to do more than one PR upstream.)
> 
> You mean, you want to run presubmit tests against the version of a pub package on some random branch on the internet? You should be able to do `dependency_overrides` with a git dependency in a pubspec.yaml. You may get an analyzer error, but I would expect everything else to Just Work:tm:

👍 
I also had to make some hacks in `packages/flutter_tools/lib/src/commands/update_packages.dart`.
And I had to use the googlesource mirrors instead of github. (We have some allowed hosts list somewhere in the infra but I couldn't find out where.)

Anyway, the CI then looked green enough to hit the publish button on the upstream changes, so the PR now only uses published deps.

**Comment by christopherfujino** [comment]
> > > @christopherfujino Is there a way to allow arbitrary git dependencies on the CI here? (If I make a PR upstream to support a PR here, having to merge and publish that upstream dependency before being able to run the CI here means its likely something will be forgotten upstream and we end up having to do more than one PR upstream.)
> > 
> > 
> > You mean, you want to run presubmit tests against the version of a pub package on some random branch on the internet? You should be able to do `dependency_overrides` with a git dependency in a pubspec.yaml. You may get an analyzer error, but I would expect everything else to Just Work:tm:
> 
> 👍 I also had to make some hacks in `packages/flutter_tools/lib/src/commands/update_packages.dart`. And I had to use the googlesource mirrors instead of github. (We have some allowed hosts list somewhere in the infra but I couldn't find out where.)
> 
> Anyway, the CI then looked green enough to hit the publish button on the upstream changes, so the PR now only uses published deps.

🤌

**Comment by dcharkes** [comment]
All green, ready for review @christopherfujino.
lmk if you want to chat about it.

**Comment by christopherfujino** [comment]
cc @andrewkolos

**Comment by christopherfujino** [comment]
Sorry for the delay @dcharkes, I forgot about this. I will take a look today, and if I have outstanding questions I will schedule a 1:1.

File: packages/flutter_tools/lib/src/isolated/native_assets/android/native_assets.dart [hunk 1]
```diff
@@ -50,14 +50,25 @@ Future<Iterable<KernelAsset>> dryRunNativeAssetsAndroidInternal(
...
+  final BuildDryRunResult buildDryRunResult = await buildRunner.buildDryRun(
     linkModePreference: LinkModePreferenceImpl.dynamic,
     targetOS: targetOS,
     workingDirectory: projectUri,
     includeParentEnvironment: true,
   );
-  ensureNativeAssetsBuildSucceed(dryRunResult);
-  final List<AssetImpl> nativeAssets = dryRunResult.assets;
+  ensureNativeAssetsBuildDryRunSucceed(buildDryRunResult);
+  final LinkDryRunResult linkDryRunResult = await buildRunner.linkDryRun(
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/isolated/native_assets/android/native_assets.dart:60]: why do we need to dry run the link phase?

**Review by christopherfujino — APPROVED** [review]
Overall LGTM. I would appreciate a refresher on why the dry runs are needed in general, and specifically for the link phase.

**Comment by dcharkes** [comment]
> Overall LGTM. I would appreciate a refresher on why the dry runs are needed in general, and specifically for the link phase.

### Why dry run in `flutter run`?

* `flutter run` compiles native assets in dry run has the kernel file with a native assets mapping sitting around unused until the first hot restart. (The native assets mapping is coming from a dry run, it doesn't build the actual assets.)
   which invokes
  * `xcodebuild` which invokes
    *  `flutter assemble` which compiles Dart to kernel with a native assets mapping and pushes it to the device. This kernel file with native assets mapping is used for hot reloads until the first hot restart. (This native assets mapping is coming from a normal run and will build the native assets.)

Because we want to support building a flutter app from the xcode project, the native assets mapping is needed in `flutter assemble`.
Because we want to support hot restart, the native assets mapping is also needed in `flutter run`.

The reason that `flutter run` does not simply do a native assets build for all target architectures of the current OS is that we would have to block until those builds are done before invoking `xcodebuild`. Since build hooks and link hooks cannot be invoked for the same configuration concurrently. Dry run is a different configuration then wet run, so in the current set up the dry-runs in `flutter run` are fired off concurrently to the `xcodebuild` process call. The fact that we do a kernel compilation in `flutter run` eagerly instead of on the first hot restart or hot reload request is to make the first hot restart or hot reload fast.

### Why run link hooks in dry run?

The build hooks can produce temporary assets (for example a static library) that are linked in a link hook (for example a dynamic library). So in order to embed the correct native assets mapping, we need the output from both hooks. (In dry run both hooks are not expected to actually produce any files, just the information about which files will be produced in a non-dry-run.)

File: packages/flutter_tools/lib/src/isolated/native_assets/android/native_assets.dart (hunk 1 above)
Comment (dcharkes) [reply packages/flutter_tools/lib/src/isolated/native_assets/android/native_assets.dart:60]: https://github.com/flutter/flutter/pull/148474#issuecomment-2125149031

**Comment by dcharkes** [comment]
Thanks @christopherfujino!

**Pull request merged**